Trouble building with cuda

Hi guys,

Sorry this is more of a support request than an issue, but hopefully the fix will be searchable for other people. When building coaster-nn with cuda feature enabled I get these errors

```
juice/coaster-nn on  master [!] is 📦 v0.4.2 via 🦀 v1.45.1 
❯ cargo build --release --no-default-features --features cuda
warning: Running bindgen(cudnn-sys), make sure to have all required host libs installed!
warning: Running bindgen(cublas-sys), make sure to have all required host libs installed!
   Compiling coaster-nn v0.4.2 (/home/matiu/projects/downloads/juice/coaster-nn)
error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_ENABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:765:66
    |
765 |             RnnPaddingMode::Enabled => Ok(cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_ENABLED),
    |                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_DISABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:766:67
    |
766 |             RnnPaddingMode::Disabled => Ok(cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_DISABLED)
    |                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_ENABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:772:36
    |
772 |             cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_ENABLED => RnnPaddingMode::Enabled,
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_DISABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:773:36
    |
773 |             cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_DISABLED => RnnPaddingMode::Disabled,
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0599`.
error: could not compile `coaster-nn`.

To learn more, run the command again with --verbose.
```
I'm on ubuntu, and I have installed `libcudnn8-dev_8.0.2.39-1+cuda11.0_amd64.deb` and I exported the paths to the libraries:

```
❯ set | rg -i 'cud'
CUBLAS_INCLUDE_DIR=/usr/local/cuda-11.0/include
CUBLAS_LIB_DIR=/usr/local/cuda-11.0/lib64
CUDNN_INCLUDE_DIR=/usr/include/x86_64-linux-gnu
CUDNN_LIB_DIR=/usr/lib/x86_64-linux-gnu
```

Any help appreciated.

It is found in one place in the include files:

```
❯ rg -C4 -n CUDNN_RNN_PADDED_IO_ENABLED /usr/local/cuda-11.0/include /usr/local/cuda-11.0/lib64 /usr/include/x86_64-linux-gnu /usr/lib/x86_64-linux-gnu
/usr/include/x86_64-linux-gnu/cudnn_adv_infer_v8.h
120-typedef unsigned cudnnRNNPaddingMode_t;
121-
122-/* For auxFlags in cudnnSetRNNDescriptor_v8() and cudnnSetRNNPaddingMode() */
123-#define CUDNN_RNN_PADDED_IO_DISABLED 0
124:#define CUDNN_RNN_PADDED_IO_ENABLED (1U << 0)
125-
126-struct cudnnRNNStruct;
127-typedef struct cudnnRNNStruct *cudnnRNNDescriptor_t;
128-
```

My issue may be because I had bits of cuda 10.1 still installed, plus cuda 11. I removed these packages:

```
❯ apt list --installed | rg cuda.*10.1
libcudart10.1/focal,now 10.1.243-3 amd64 [installed,automatic]
nvidia-cuda-dev/focal,now 10.1.243-3 amd64 [installed]
nvidia-cuda-doc/focal,focal,now 10.1.243-3 all [installed,automatic]
nvidia-cuda-gdb/focal,now 10.1.243-3 amd64 [installed,automatic]
nvidia-cuda-toolkit/focal,now 10.1.243-3 amd64 [installed]
```

I also found that I had to append a `/` to my include directories because they were symlinks set up by the ubuntu packages.

Now I'm at this point

```
juice/coaster-nn on  master [!?] is 📦 v0.4.2 via 🦀 v1.45.1 
❯ export CUBLAS_INCLUDE_DIR=/usr/local/cuda-11.0/include/

juice/coaster-nn on  master [!?] is 📦 v0.4.2 via 🦀 v1.45.1 
❯ export CUBLAS_LIB_DIR=/usr/local/cuda-11.0/lib64/

juice/coaster-nn on  master [!?] is 📦 v0.4.2 via 🦀 v1.45.1 
❯ export CUDNN_INCLUDE_DIR=/usr/include/x86_64-linux-gnu/

juice/coaster-nn on  master [!?] is 📦 v0.4.2 via 🦀 v1.45.1 
❯ export CUDNN_LIB_DIR=/usr/lib/x86_64-linux-gnu/

juice/coaster-nn on  master [!?] is 📦 v0.4.2 via 🦀 v1.45.1 
❯ cargo build --release --no-default-features --features cuda
   Compiling rcudnn-sys v0.4.0 (/home/matiu/projects/downloads/juice/rcudnn/cudnn-sys)
warning: Running bindgen(cublas-sys), make sure to have all required host libs installed!
The following warnings were emitted during compilation:

warning: Running bindgen(cudnn-sys), make sure to have all required host libs installed!

error: failed to run custom build command for `rcudnn-sys v0.4.0 (/home/matiu/projects/downloads/juice/rcudnn/cudnn-sys)`

Caused by:
  process didn't exit successfully: `/home/matiu/projects/downloads/juice/target/release/build/rcudnn-sys-27813d302dbee00b/build-script-build` (exit code: 101)
--- stdout
cargo:rustc-link-search=native=/usr/lib/x86_64-linux-gnu/
cargo:rustc-link-lib=dylib=cudnn
cargo:rustc-link-lib=dylib=cudart
cargo:rustc-link-lib=dylib=cuda
cargo:include=/usr/include/x86_64-linux-gnu/
cargo:rerun-if-changed=build.rs
cargo:warning=Running bindgen(cudnn-sys), make sure to have all required host libs installed!

--- stderr
/usr/include/cudnn.h:57:10: fatal error: 'cuda_runtime.h' file not found
/usr/include/cudnn.h:57:10: fatal error: 'cuda_runtime.h' file not found, err: true
thread 'main' panicked at 'Unable to generate bindings: ()', rcudnn/cudnn-sys/build.rs:67:24
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Specifically it can't find `cuda_runtime.h` but it is included in my environment  variables:

```
❯ find $CUBLAS_INCLUDE_DIR   -type f -name cuda_runtime.h
/usr/local/cuda-11.0/include/cuda_runtime.h
```

I'm still wrestling. I'll update here more if I figure it out.

It seems to be ignoring my `CUBLAS_INCLUDE_DIR` so I did a terrible hack and copied all the include files to `/usr/local/all_cuda_includes` and passed that as my `CUDNN_INCLUDE_DIR`

Now I'm back to the same place; the `CUDNN_RNN_PADDED_IO_ENABLED` exists in the include files:

```
❯ rg -C4 CUDNN_RNN_PADDED_IO_ENABLED /usr/local/all_cuda_includes/
/usr/local/all_cuda_includes/cudnn_adv_infer_v8.h
120-typedef unsigned cudnnRNNPaddingMode_t;
121-
122-/* For auxFlags in cudnnSetRNNDescriptor_v8() and cudnnSetRNNPaddingMode() */
123-#define CUDNN_RNN_PADDED_IO_DISABLED 0
124:#define CUDNN_RNN_PADDED_IO_ENABLED (1U << 0)
125-
126-struct cudnnRNNStruct;
127-typedef struct cudnnRNNStruct *cudnnRNNDescriptor_t;
128-
```

But the same error prevails when trying to compile:

```
❯ RUST_BACKTRACE=full cargo build --release --no-default-features --features cuda
warning: Running bindgen(cudnn-sys), make sure to have all required host libs installed!
warning: Running bindgen(cublas-sys), make sure to have all required host libs installed!
   Compiling coaster-nn v0.4.2 (/home/matiu/projects/downloads/juice/coaster-nn)
error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_ENABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:765:66
    |
765 |             RnnPaddingMode::Enabled => Ok(cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_ENABLED),
    |                                                                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_DISABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:766:67
    |
766 |             RnnPaddingMode::Disabled => Ok(cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_DISABLED)
    |                                                                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_ENABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:772:36
    |
772 |             cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_ENABLED => RnnPaddingMode::Enabled,
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error[E0599]: no associated item named `CUDNN_RNN_PADDED_IO_DISABLED` found for type `u32` in the current scope
   --> coaster-nn/src/frameworks/cuda/mod.rs:773:36
    |
773 |             cudnnRNNPaddingMode_t::CUDNN_RNN_PADDED_IO_DISABLED => RnnPaddingMode::Disabled,
    |                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ associated item not found in `u32`

error: aborting due to 4 previous errors

For more information about this error, try `rustc --explain E0599`.
error: could not compile `coaster-nn`.

To learn more, run the command again with --verbose.
```

All help and direction appreciated :)

--- a/coaster-blas/Cargo.toml
+++ b/coaster-blas/Cargo.toml
@@ -2,7 +2,6 @@
 name = "coaster-blas"
 description = "Coaster library for full BLAS support"
 version = "0.3.0"
-edition = "2018"
 authors = [
     "Bernhard Schuster <bernhard@ahoi.io>",
     "Lissa Hyacinth <lissahyacinth@gmail.com>",
@@ -16,6 +15,7 @@ documentation = "https://spearow.github.com/coaster-blas"
 readme = "README.md"
 keywords = ["blas", "coaster", "computation", "hpc", "plugin"]
 license = "MIT OR Apache-2.0"
+edition = "2018"
 
 [dependencies]
 coaster = { version = "0.1", path = "../coaster", default-features = false }

--- a/coaster-nn/Cargo.toml
+++ b/coaster-nn/Cargo.toml
@@ -14,6 +14,7 @@ documentation = "https://spearow.io/projects/coaster-nn/documentation"
 readme = "README.md"
 keywords = ["neural-network", "coaster", "computation", "hpc", "plugin"]
 license = "MIT OR Apache-2.0"
+edition = "2018"
 
 [dependencies]
 coaster = { version = "0.1", default-features = false }

--- a/coaster-nn/src/frameworks/native/helper.rs
+++ b/coaster-nn/src/frameworks/native/helper.rs
@@ -5,7 +5,7 @@ use crate::co::plugin::Error as PluginError;
 use crate::co::plugin::numeric_helpers::Float;
 use crate::co::frameworks::native::flatbox::FlatBox;
 use crate::{DirectionMode, RnnInputMode};
-use RnnNetworkMode;
+use crate::RnnNetworkMode;
 
 #[derive(Debug, Copy, Clone)]
 #[allow(missing_docs)]

--- a/coaster-nn/src/lib.rs
+++ b/coaster-nn/src/lib.rs
@@ -66,7 +66,7 @@
 //! #
 //! # fn main() {
 //! #     if cfg!(feature = "cuda") {
-//! #         ::cuda::main();
+//! #         cuda::main();
 //! #    }
 //! # }
 //! ```

--- a/coaster/Cargo.toml
+++ b/coaster/Cargo.toml
@@ -8,7 +8,7 @@ authors = [
     "Michael Hirn",
     "Maximilian Goisser",
 ]
-
+edition = "2018"
 repository = "https://github.com/spearow/coaster"
 homepage = "https://spearow.io/projects/coaster"
 documentation = "https://spearow.github.io/coaster"

--- a/coaster/examples/readme.rs
+++ b/coaster/examples/readme.rs
@@ -3,6 +3,7 @@ use coaster_nn as nn;
 #[cfg(feature = "cuda")]
 use rcublas;
 
+use nn::;
 use co::prelude::*;
 use co::frameworks::native::flatbox::FlatBox;
 #[cfg(feature = "cuda")]
@@ -25,11 +26,11 @@ fn main() {
     let backend = get_native_backend();
 
     // Initialize two SharedTensors.
-    let mut x = SharedTensor::<f32>::new(&(1, 1, 3));
-    let result = SharedTensor::<f32>::new(&(1, 1, 3));
+    let mut x = nn::SharedTensor::<f32>::new(&(1, 1, 3));
+    let result = nn::SharedTensor::<f32>::new(&(1, 1, 3));
     // Fill `x` with some data.
     let payload: &[f32] = &::std::iter::repeat(1f32).take(x.capacity()).collect::<Vec<f32>>();
-    let native = Backend::<Native>::default().unwrap();
+    let native = nn::Backend::<Native>::default().unwrap();
     write_to_memory(x.write_only(native.device()).unwrap(), payload); // Write to native host memory.
     // Run the sigmoid operation, provided by the NN Plugin, on your CUDA enabled GPU.
     // FIXME: Sigmoid cannot be included from coaster-nn without using cuda and native features

--- a/coaster/examples/readme.rs
+++ b/coaster/examples/readme.rs
@@ -26,17 +26,17 @@ fn main() {
     let backend = get_native_backend();
 
     // Initialize two SharedTensors.
-    let mut x = nn::SharedTensor::<f32>::new(&(1, 1, 3));
-    let result = nn::SharedTensor::<f32>::new(&(1, 1, 3));
+    let mut x = SharedTensor::<f32>::new(&(1, 1, 3));
+    // let mut result = SharedTensor::<f32>::new(&(1, 1, 3));
     // Fill `x` with some data.
     let payload: &[f32] = &::std::iter::repeat(1f32).take(x.capacity()).collect::<Vec<f32>>();
-    let native = nn::Backend::<Native>::default().unwrap();
+    let native = Backend::<Native>::default().unwrap();
     write_to_memory(x.write_only(native.device()).unwrap(), payload); // Write to native host memory.
     // Run the sigmoid operation, provided by the NN Plugin, on your CUDA enabled GPU.
     // FIXME: Sigmoid cannot be included from coaster-nn without using cuda and native features
     // from coaster-nn. This causes the error https://github.com/rust-lang/cargo/issues/6915 ,
     // and so sigmoid has been disabled for now.
-    backend.sigmoid(&mut x, &mut result).unwrap();
+    // backend.sigmoid(&mut x, &mut result).unwrap();
     // See the result.
-    println!("{:?}", result.read(native.device()).unwrap().as_slice::<f32>());
+    // println!("{:?}", result.read(native.device()).unwrap().as_slice::<f32>());
 }

--- a/coaster/examples/readme.rs
+++ b/coaster/examples/readme.rs
@@ -3,7 +3,7 @@ use coaster_nn as nn;
 #[cfg(feature = "cuda")]
 use rcublas;
 
-use nn::;
+use nn::*;
 use co::prelude::*;
 use co::frameworks::native::flatbox::FlatBox;
 #[cfg(feature = "cuda")]

--- a/coaster/examples/readme.rs
+++ b/coaster/examples/readme.rs
@@ -1,14 +1,13 @@
-extern crate coaster as co;
-extern crate coaster_nn as nn;
+use coaster as co;
+use coaster_nn as nn;
 #[cfg(feature = "cuda")]
-extern crate rcublas;
+use rcublas;
 
 use co::prelude::*;
 use co::frameworks::native::flatbox::FlatBox;
-use nn::*;
 #[cfg(feature = "cuda")]
 use co::frameworks::cuda::get_cuda_backend;
-#[cfg(feature = "native")]
+#[cfg(not(feature = "cuda"))]
 use co::frameworks::native::get_native_backend;
 
 fn write_to_memory<T: Copy>(mem: &mut FlatBox, data: &[T]) {
@@ -20,14 +19,14 @@ fn write_to_memory<T: Copy>(mem: &mut FlatBox, data: &[T]) {
 
 fn main() {
     #[cfg(feature = "cuda")]
-        let backend = get_cuda_backend();
+    let backend = get_cuda_backend();
 
     #[cfg(not(feature = "cuda"))]
-        let backend = get_native_backend();
+    let backend = get_native_backend();
 
     // Initialize two SharedTensors.
     let mut x = SharedTensor::<f32>::new(&(1, 1, 3));
-    let mut result = SharedTensor::<f32>::new(&(1, 1, 3));
+    let result = SharedTensor::<f32>::new(&(1, 1, 3));
     // Fill `x` with some data.
     let payload: &[f32] = &::std::iter::repeat(1f32).take(x.capacity()).collect::<Vec<f32>>();
     let native = Backend::<Native>::default().unwrap();
@@ -36,8 +35,7 @@ fn main() {
     // FIXME: Sigmoid cannot be included from coaster-nn without using cuda and native features
     // from coaster-nn. This causes the error https://github.com/rust-lang/cargo/issues/6915 ,
     // and so sigmoid has been disabled for now.
-    //backend.sigmoid(&mut x, &mut result).unwrap();
+    backend.sigmoid(&mut x, &mut result).unwrap();
     // See the result.
     println!("{:?}", result.read(native.device()).unwrap().as_slice::<f32>());
 }
-

--- a/coaster/src/frameworks/cuda/mod.rs
+++ b/coaster/src/frameworks/cuda/mod.rs
@@ -19,7 +19,7 @@ pub use self::device::{Device, DeviceInfo};
 pub use self::api::{Driver, DriverError};
 use crate::cudnn::*;
 use crate::cublas;
-use BackendConfig;
+use crate::BackendConfig;
 
 pub mod device;
 pub mod context;

--- a/coaster/src/frameworks/native/mod.rs
+++ b/coaster/src/frameworks/native/mod.rs
@@ -58,12 +58,10 @@ impl IFramework for Native {
     fn ID() -> &'static str { "NATIVE" }
 
     fn new() -> Native {
-        match Native::load_hardwares() {
-            Ok(hardwares) => Native {
-                hardwares,
-                binary: Binary::new(),
-            },
-            Err(err) => panic!(err)
+        let hardwares = Native::load_hardwares().expect("Native hardwares are always ok. qed");
+        Self {
+            hardwares,
+            binary: Binary::new(),
         }
     }
 

--- a/coaster/src/frameworks/opencl/mod.rs
+++ b/coaster/src/frameworks/opencl/mod.rs
@@ -51,14 +51,10 @@ impl IFramework for OpenCL {
     fn ID() -> &'static str { "OPENCL" }
 
     fn new() -> OpenCL {
-        match OpenCL::load_hardwares() {
-            Ok(hardwares) => {
-                OpenCL {
-                    hardwares: hardwares,
-                    binary: Program::from_isize(1)
-                }
-            },
-            Err(err) => panic!(err)
+        let hardwares = OpenCL::load_hardwares().expect("Acquiring hw never fails. qed");
+        Self {
+            hardwares,
+            binary: Program::from_isize(1)
         }
     }
 

--- a/coaster/tests/shared_memory_specs.rs
+++ b/coaster/tests/shared_memory_specs.rs
@@ -7,6 +7,7 @@ mod shared_memory_spec {
     use crate::co::tensor::Error;
     use crate::co::frameworks::native::flatbox::FlatBox;
 
+    #[cfg(features = "cuda")]
     fn write_to_memory<T: Copy>(mem: &mut FlatBox, data: &[T]) {
         let mem_buffer = mem.as_mut_slice::<T>();
         for (index, datum) in data.iter().enumerate() {

--- a/coaster/tests/shared_memory_specs.rs
+++ b/coaster/tests/shared_memory_specs.rs
@@ -3,9 +3,9 @@ use libc;
 
 #[cfg(test)]
 mod shared_memory_spec {
-    use crate::co::prelude::*;
-    use crate::co::tensor::Error;
-    use crate::co::frameworks::native::flatbox::FlatBox;
+    use super::co::prelude::*;
+    use super::co::tensor::Error;
+    use super::co::frameworks::native::flatbox::FlatBox;
 
     #[cfg(features = "cuda")]
     fn write_to_memory<T: Copy>(mem: &mut FlatBox, data: &[T]) {

--- a/coaster/tests/shared_memory_specs.rs
+++ b/coaster/tests/shared_memory_specs.rs
@@ -1,5 +1,5 @@
-extern crate coaster as co;
-extern crate libc;
+use coaster as co;
+use libc;
 
 #[cfg(test)]
 mod shared_memory_spec {

--- a/greenglas/Cargo.toml
+++ b/greenglas/Cargo.toml
@@ -10,6 +10,7 @@ documentation = "https://spearow.github.com/greenglas"
 readme = "README.md"
 keywords = ["preprocessing", "data-preprocessing", "transformation", "encoding", "machine-learning"]
 license = "MIT OR Apache-2.0"
+edition = "2018"
 
 [dependencies]
 coaster = { version = "0.1", default-features = false, features = ["native"] } # native feature to read/write data into tensors

--- a/juice-examples/juice-utils/Cargo.toml
+++ b/juice-examples/juice-utils/Cargo.toml
@@ -2,35 +2,11 @@
 name = "juice-utils"
 description = "Utilities for running Juice Examples"
 version = "0.0.1"
-authors = ["Lissa Hyacinth <lissahyacinth@gmail.com>"]
+authors = ["Lissa Hyacinth <lissahyacinth@gmail.com>", "Bernhard Schuster <bernhard@spearow.io>"]
 edition = "2018"
 
 [dependencies]
-greenglas = { version = "0.2" }
-juice = { version = "0.2.4", default-features = false }
-coaster = { version = "0.1.1", default-features = false }
-coaster-nn = { version = "0.4.2", default-features = false }
-
-csv = "1"
-reqwest= {version = "0.10.6", features = ["blocking"]}
-futures = "0.3.5"
-futures-util = "0.3"
-
-tokio = {version = "0.2.4", features = ["macros"]}
-bytes="*"
+reqwest= {version = "0.11.2", features = ["blocking"]}
 flate2 = "1"
-mnist = "0.4.0"
-
+fs-err = "2.5.0"
 log = "0.4"
-env_logger = "0.7"
-timeit = "0.1.2"
-
-docopt = "1"
-serde = { version = "1", features = ["derive"] }
-
-
-[features]
-default = ["native", "cuda"]
-native = ["juice/native", "coaster/native"]
-opencl = ["juice/opencl", "coaster/opencl"]
-cuda = ["juice/cuda", "coaster/cuda"]

--- a/juice-examples/juice-utils/src/lib.rs
+++ b/juice-examples/juice-utils/src/lib.rs
@@ -9,18 +9,19 @@ pub fn download_datasets(datasets: &[&str], asset_path: &str, base_url: &str) ->
     std::fs::create_dir_all(asset_path)?;
     for dataset in datasets {
         let url = format!("{}/{}", base_url, dataset);
+        log::info!("Downloading {}", dataset);
         let resp = client.get(&url).send()?.bytes()?;
-        println!("Downloading {}", dataset);
         let name = format!("{}/{}", asset_path, dataset);
         let mut f = fs::File::create(name.clone()).expect("Failed to create file");
-        f.write_all(&resp).unwrap();
+        f.write_all(&resp)?;
+        log::info!("Download of {} complete.", dataset);
     }
     Ok(())
 }
 
 pub fn unzip_datasets(datasets: &[&str], asset_path: &str) -> io::Result<()> {
     for filename in datasets {
-        println!("Decompressing: {}", filename);
+        log::info!("Decompressing: {}", filename);
 
         let file_handle = fs::File::open(&format!("{}/{}", asset_path, filename)).unwrap();
         let mut decoder = GzDecoder::new(file_handle);
@@ -29,6 +30,7 @@ pub fn unzip_datasets(datasets: &[&str], asset_path: &str) -> io::Result<()> {
 
         let mut dest = fs::File::create(format!("{}/{}", asset_path, filename_string))?;
         std::io::copy(&mut decoder, &mut dest)?;
+        log::info!("Decompression of {} complete.", filename);
     }
     Ok(())
 }

--- a/juice-examples/juice-utils/src/lib.rs
+++ b/juice-examples/juice-utils/src/lib.rs
@@ -1,12 +1,8 @@
-extern crate bytes;
-extern crate flate2;
-
 use flate2::read::GzDecoder;
-use reqwest;
 use reqwest::blocking::Client;
-use std::fs::File;
+use fs_err as fs;
 use std::io::prelude::*;
-
+use std::io;
 
 pub fn download_datasets(datasets: &[&str], asset_path: &str, base_url: &str) -> Result<(), Box<dyn std::error::Error>> {
     let client = Client::new();
@@ -16,31 +12,23 @@ pub fn download_datasets(datasets: &[&str], asset_path: &str, base_url: &str) ->
         let resp = client.get(&url).send()?.bytes()?;
         println!("Downloading {}", dataset);
         let name = format!("{}/{}", asset_path, dataset);
-        let mut f = File::create(name.clone()).expect("Failed to create file");
-        f.write(&resp).unwrap();
+        let mut f = fs::File::create(name.clone()).expect("Failed to create file");
+        f.write_all(&resp).unwrap();
     }
     Ok(())
 }
 
-pub fn unzip_datasets(datasets: &[&str], asset_path: &str) {
+pub fn unzip_datasets(datasets: &[&str], asset_path: &str) -> io::Result<()> {
     for filename in datasets {
         println!("Decompressing: {}", filename);
 
-        let mut file_handle = File::open(&format!("{}/{}", asset_path, filename)).unwrap();
-        let mut in_file: Vec<u8> = Vec::new();
-        let mut decompressed_file: Vec<u8> = Vec::new();
-
-        file_handle.read_to_end(&mut in_file).unwrap();
-
-        let mut decoder = GzDecoder::new(in_file.as_slice());
-
-        decoder.read_to_end(&mut decompressed_file).unwrap();
+        let file_handle = fs::File::open(&format!("{}/{}", asset_path, filename)).unwrap();
+        let mut decoder = GzDecoder::new(file_handle);
 
         let filename_string = filename.split(".").nth(0).unwrap();
 
-        File::create(format!("{}/{}", asset_path, filename_string))
-            .unwrap()
-            .write_all(&decompressed_file as &[u8])
-            .unwrap();
+        let mut dest = fs::File::create(format!("{}/{}", asset_path, filename_string))?;
+        std::io::copy(&mut decoder, &mut dest)?;
     }
+    Ok(())
 }

--- a/juice-examples/mackey-glass-rnn-regression/Cargo.toml
+++ b/juice-examples/mackey-glass-rnn-regression/Cargo.toml
@@ -13,7 +13,8 @@ coaster-nn = { version = "0.4.2", default-features = false }
 
 csv = "1"
 serde = { version = "1", features = ["derive"] }
-env_logger = "0.7.1"
+env_logger = "0.8.3"
+log = "0.4.14"
 docopt= "1.1.0"
 
 [features]

--- a/juice-examples/mackey-glass-rnn-regression/src/main.rs
+++ b/juice-examples/mackey-glass-rnn-regression/src/main.rs
@@ -24,11 +24,11 @@ Usage:  mackey-glass-example train [--file=<path>] [--batchSize=<batch>] [--lear
         mackey-glass-example test [--file=<path>]
 
 Options:
-    --file=<path>  Filepath for saving trained network.
+    --file=<path>        Filepath for saving trained network.
     --batchSize=<batch>  Network Batch Size.
     --learningRate=<lr>  Learning Rate.
-    --momentum=<float>  Momentum.
-    -h, --help  Show this screen.
+    --momentum=<float>   Momentum.
+    -h, --help           Show this screen.
 ";
 
 #[allow(non_snake_case)]
@@ -49,17 +49,23 @@ enum DataMode {
     Test,
 }
 
+impl DataMode {
+    fn as_path(&self) -> &'static str {
+        match self {
+            DataMode::Train => "assets/norm_mackeyglass_train.csv",
+            DataMode::Test => "assets/norm_mackeyglass_test.csv",
+        }
+    }
+}
+
 const TRAIN_ROWS: usize = 35192;
 const TEST_ROWS: usize = 8798;
 const DATA_COLUMNS: usize = 10;
 
 // Provide an Iterator over the input data
 fn data_generator(data: DataMode) -> impl Iterator<Item = (f32, Vec<f32>)> {
-    let path = match data {
-        DataMode::Train => "assets/norm_mackeyglass_train.csv",
-        DataMode::Test => "assets/norm_mackeyglass_test.csv",
-    };
-    let rdr = Reader::from_reader(File::open(path).unwrap());
+
+    let rdr = Reader::from_reader(File::open(data.as_path()).unwrap());
     rdr.into_deserialize().map(move |row| match row {
         Ok(value) => {
             let row_vec: Box<Vec<f32>> = Box::new(value);
@@ -68,7 +74,7 @@ fn data_generator(data: DataMode) -> impl Iterator<Item = (f32, Vec<f32>)> {
             (label, columns)
         }
         _ => {
-            println!("no value");
+            log::error!("file seems to be empty");
             panic!();
         }
     })

--- a/juice-examples/mnist-image-multiclass-classification/src/main.rs
+++ b/juice-examples/mnist-image-multiclass-classification/src/main.rs
@@ -5,8 +5,9 @@ extern crate juice;
 extern crate juice_utils;
 extern crate mnist;
 
-#[cfg(all(feature = "cuda"))]
+#[cfg(feature = "cuda")]
 use co::frameworks::cuda::get_cuda_backend;
+#[cfg(not(feature = "cuda"))]
 use co::frameworks::native::get_native_backend;
 use co::prelude::*;
 use juice::layer::*;

--- a/juice-examples/mnist-image-multiclass-classification/src/main.rs
+++ b/juice-examples/mnist-image-multiclass-classification/src/main.rs
@@ -79,7 +79,7 @@ fn main() {
                     &"./assets/mnist/",
                     "http://yann.lecun.com/exdb/mnist/").unwrap();
 
-                unzip_datasets(&datasets, &"./assets/mnist/");
+                unzip_datasets(&datasets, &"./assets/mnist/").unwrap();
             }
 
             "fashion" => {
@@ -97,7 +97,7 @@ fn main() {
                 ).unwrap();
                 println!("{}", "Fashion MNIST dataset downloaded".to_string());
 
-                unzip_datasets(&datasets, &"./assets/mnist-fashion/");
+                unzip_datasets(&datasets, &"./assets/mnist-fashion/").unwrap();
                 println!("{}", "Fashion MNIST dataset decompressed".to_string());
             }
             _ => println!("{}", "Failed to download MNIST dataset!".to_string()),

--- a/juice/tests/layer_specs.rs
+++ b/juice/tests/layer_specs.rs
@@ -7,6 +7,7 @@ mod layer_spec {
     use juice::layer::*;
     use juice::weight::{DimCheckMode, WeightConfig};
     use std::rc::Rc;
+
     // only used by cuda right now
     #[allow(dead_code)]
     fn new_layer_config() -> LayerConfig {
@@ -119,6 +120,76 @@ mod layer_spec {
 
     #[cfg(feature = "cuda")]
     mod cuda {
+        macro_rules! assert_slice_eq {
+            ($lv:expr, $rv:expr, $eps:expr, $($arg:tt)+) => {
+                {
+                    let eps = $eps;
+                    let lv = $lv;
+                    let rv = $rv;
+                    for (i,(a,b)) in lv.iter().zip(rv.iter()).enumerate() {
+                        let delta = if *a < *b {
+                            *a - *b
+                        } else if *b < *a {
+                            *b - *a
+                        } else {
+                            continue;
+                        };
+                        if delta > eps {
+                            ::std::panic!(r#"assertion failed: `(left == right)`
+                            left: `{:?}`,
+                           right: `{:?}`: since value {} at index {} did not match {} ± {}: {}"#,
+                           &lv, &rv, a, i, b, $eps,
+                           ::std::format_args!($($arg)+))
+                        }
+                    }
+                    assert_eq!(lv.iter().count(), rv.iter().count(), "Left and right handside have matching prefix, but not length.");
+                }
+            };
+
+            ($lv:expr, $rv:expr, $eps:expr) => {
+                {
+                    let eps = $eps;
+                    let lv = $lv;
+                    let rv = $rv;
+                    for (i,(a,b)) in lv.iter().zip(rv.iter()).enumerate() {
+                        let delta = if *a < *b {
+                            *b - *a
+                        } else if *b < *a {
+                            *a - *b
+                        } else {
+                            continue;
+                        };
+                        if delta > eps {
+                            ::std::panic!(r#"assertion failed: `(left == right)`
+                            left: `{:?}`,
+                           right: `{:?}`: since value {} at index {} did not match {} ± {}"#,
+                           &lv, &rv, a, i, b, $eps);
+                        }
+                    }
+                    assert_eq!(lv.iter().count(), rv.iter().count(), "Left and right handside have matching prefix, but not length.");
+                }
+            };
+        }
+
+        #[test]
+        fn macro_test_assert_slice_eq() {
+            assert_slice_eq!(&[0.51], &[0.52], 0.19999999);
+            assert_slice_eq!(&[0.51], &[0.51], 0.00000001);
+        }
+
+
+        #[test]
+        #[should_panic]
+        fn macro_test_assert_slice_eq_not() {
+            assert_slice_eq!(&[0.51], &[0.52], 0.00000001);
+        }
+
+        #[test]
+        #[should_panic]
+        fn macro_test_assert_slice_eq_len() {
+            assert_slice_eq!(&[0.50, 0.50], &[0.50], 0.00000001);
+        }
+
         use super::{cuda_backend, native_backend};
         use crate::co::prelude::*;
         use juice::layer::*;
@@ -211,21 +282,24 @@ mod layer_spec {
             let normal_tensor_output = normal_network.forward(&[Arc::new(RwLock::new(normal_tensor))])[0].clone();
             let normal_tensor_output_native_ = normal_tensor_output.read().unwrap();
             let normal_tensor_output_native = normal_tensor_output_native_.read(native_backend.device()).unwrap();
-            assert_eq!(
+            assert_slice_eq!(
                 &[0.7310585786f32, 0.7310586f32, 0.880797f32],
-                normal_tensor_output_native.as_slice::<f32>()
+                normal_tensor_output_native.as_slice::<f32>(),
+                1e-6_f32
             );
 
             let reshape_tensor_output = reshape_network.forward(&[Arc::new(RwLock::new(reshape_tensor))])[0].clone();
             let reshape_tensor_output_native_ = reshape_tensor_output.read().unwrap();
             let reshape_tensor_output_native = reshape_tensor_output_native_.read(native_backend.device()).unwrap();
-            assert_eq!(
+            assert_slice_eq!(
                 &[0.7310585786f32, 0.7310586f32, 0.880797f32],
-                reshape_tensor_output_native.as_slice::<f32>()
+                reshape_tensor_output_native.as_slice::<f32>(),
+                1e-6_f32
             );
-            assert_eq!(
+            assert_slice_eq!(
                 normal_tensor_output_native.as_slice::<f32>(),
-                reshape_tensor_output_native.as_slice::<f32>()
+                reshape_tensor_output_native.as_slice::<f32>(),
+                1e-6_f32
             );
         }
     }

--- a/rcudnn/cudnn/Cargo.toml
+++ b/rcudnn/cudnn/Cargo.toml
@@ -2,7 +2,7 @@
 name = "rcudnn"
 description = "safe Rust wrapper for CUDA's cuDNN"
 version = "1.7.0"
-#edition = "2018"
+edition = "2018"
 authors = [
     "Bernhard Schuster <bernhard@ahoi.io>",
     "Lissa Hyacinth <lissahyacinth@gmail.com>",

--- a/rcudnn/cudnn/src/api/rnn.rs
+++ b/rcudnn/cudnn/src/api/rnn.rs
@@ -4,7 +4,7 @@
 
 use crate::ffi::*;
 use crate::{Error, API};
-use utils::DataType;
+use crate::utils::DataType;
 
 // Workspace
 impl API {
@@ -480,7 +480,7 @@ impl API {
         cy_desc: cudnnTensorDescriptor_t,
         cy: *mut ::libc::c_void,
         work_space: *mut ::libc::c_void,
-        work_size_in_bytes: ::libc::size_t, 
+        work_size_in_bytes: ::libc::size_t,
     ) -> Result<(), Error> {
        unsafe {
            API::ffi_rnn_forward_inference(
@@ -505,7 +505,7 @@ impl API {
                work_size_in_bytes,
 
            )
-       } 
+       }
     }
     #[allow(clippy::too_many_arguments)]
     unsafe fn ffi_rnn_forward_inference(
@@ -777,7 +777,7 @@ impl API {
         dw_desc: cudnnFilterDescriptor_t,
         dw: *mut ::libc::c_void,
         reserve_space: *const ::libc::c_void,
-        reserve_space_size_in_bytes: usize, 
+        reserve_space_size_in_bytes: usize,
     ) -> Result<(), Error> {
         unsafe {
             API::ffi_rnn_backward_weights(

--- a/rcudnn/cudnn/src/api/utils.rs
+++ b/rcudnn/cudnn/src/api/utils.rs
@@ -3,7 +3,7 @@
 use crate::ffi::*;
 use std::ptr;
 use crate::{Error, API};
-use utils::DataType;
+use crate::utils::DataType;
 
 impl API {
     /// Initialize the CUDA cuDNN API with needed context and resources.

--- a/rcudnn/cudnn/src/rnn_descriptor.rs
+++ b/rcudnn/cudnn/src/rnn_descriptor.rs
@@ -3,13 +3,11 @@
 //! A Recurrent Descriptor is used to hold information about the rnn,
 //! which is needed for forward and backward rnnal operations.
 
-use super::{API, Error};
-use super::utils::DataType;
 use ffi::*;
+use super::{API, Error};
+use crate::utils::DropoutConfig;
+use crate::utils::DataType;
 use crate::Cudnn;
-
-
-use utils::DropoutConfig;
 
 /// Describes a Recurrent Descriptor.
 #[derive(Debug)]

